File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    private Map<Character, String> buttons = new HashMap<>();

    public void backtrack(String digits, List<String> answer, String temp, int i) {
        
        if (!temp.isEmpty() && temp.length() == digits.length()) {
            answer.add(temp);
        }

        if (digits.length() <= i) {
            return;
        }

 
        char digit = digits.charAt(i);
        String letter = buttons.get(digit);
        for (char c: letter.toCharArray()) {
            temp += c;
            backtrack(digits, answer, temp, i+1);
            temp = temp.substring(0, temp.length() - 1);
        }
    }
    public List<String> letterCombinations(String digits) {
        buttons.put('2', "abc");
        buttons.put('3', "def");
        buttons.put('4', "ghi");
        buttons.put('5', "jkl");
        buttons.put('6', "mno");
        buttons.put('7', "pqrs");
        buttons.put('8', "tuv");
        buttons.put('9', "wxyz");
        List<String> answer = new ArrayList<>();
        backtrack(digits, answer, "", 0);
        return answer;
    }
}

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        phone_letters = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }

        letter_combination = []

        def backtrack(digits: str, curr: int, form: str) -> None:
            if curr >= len(digits):
                return
            digit = digits[curr]
            letters = phone_letters[digit]
            for letter in letters:
                form += letter

                if len(form) == len(digits):
                    letter_combination.append(form)
                backtrack(digits, curr + 1, form)
                form = form[: len(form) - 1]

        backtrack(digits, 0, "")
        return letter_combination
